### PR TITLE
feat: Add necessary middlewares for tax inclusive pricing

### PR DESCRIPTION
**Pull request by sradevski**
We are adding tax inclusive pricing calculation when listing products.

Two things to keep in mind:
- `region_id` will be required if you request calculated prices.
- We won't accept `currency_code` anymore, as that will come from the region info (since ultimately a cart and its currency are tied to a region)

REF CORE-2376
DEPENDS ON #8003

### CODE REVIEW

File: packages/medusa/src/api/store/products/helpers.ts [hunk 1]
```diff
@@ -30,3 +47,101 @@ export const maybeApplyStockLocationId = async (req: MedusaRequest, ctx) => {
...
+}
+
+const getTaxPricesForVariant = (
+  variant: HttpTypes.StoreProductVariant,
+  taxRatesForVariant: ItemTaxLineDTO[],
+  isTaxInclusive: boolean
+) => {
+  const tax = calculateTaxTotal({
+    // TODO: See why rate is nullable in tax
+    taxLines: taxRatesForVariant as any,
```
Comment (olivermrbl) [inline packages/medusa/src/api/store/products/helpers.ts]: Can we avoid the `any` here?

File: packages/medusa/src/api/store/products/helpers.ts [hunk 2]
```diff
@@ -30,3 +47,101 @@ export const maybeApplyStockLocationId = async (req: MedusaRequest, ctx) => {
...
+      const { priceWithTax, priceWithoutTax } = getTaxPricesForVariant(
+        variant,
+        taxRatesForVariant,
+        isTaxInclusive
+      )
+
+      ;(variant.calculated_price as any).calculated_amount_with_tax =
+        priceWithTax
+      ;(variant.calculated_price as any).calculated_amount_without_tax =
+        priceWithoutTax
```
Comment (olivermrbl) [inline packages/medusa/src/api/store/products/helpers.ts:97]:
```suggestion
      ;variant.calculated_price["calculated_amount_with_tax"] =
        priceWithTax
      ;variant.calculated_price["calculated_amount_without_tax"] =
        priceWithoutTax
```

File: packages/medusa/src/api/store/products/helpers.ts [hunk 3]
```diff
@@ -30,3 +47,101 @@ export const maybeApplyStockLocationId = async (req: MedusaRequest, ctx) => {
...
+) => {
+  const tax = calculateTaxTotal({
+    // TODO: See why rate is nullable in tax
+    taxLines: taxRatesForVariant as any,
+    taxableAmount: variant.calculated_price!.calculated_amount!,
+    includesTax: isTaxInclusive,
+  })
+
+  const priceWithoutTax = isTaxInclusive
+    ? tax.minus(variant.calculated_price!.calculated_amount!).abs().toNumber()
```
Comment (olivermrbl) [inline packages/medusa/src/api/store/products/helpers.ts]: Wouldn't it be more human readable to do `price - tax` instead of `absolute(tax - price)`?

**Review by olivermrbl — DISMISSED** [review]
Overall LGTM, there are a few todos here and there, but I assume these will be tackled in follow-up PRs?

File: integration-tests/http/__tests__/product/store/product.spec.ts [hunk 4]
```diff
@@ -894,7 +894,7 @@ medusaIntegrationTestRunner({
         expect(error.response.status).toEqual(400)
         expect(error.response.data).toEqual({
           message:
-            "Missing required pricing context to calculate prices - currency_code or region_id",
+            "Missing required pricing context to calculate prices - region_id",
```
Comment (riqwan) [inline integration-tests/http/__tests__/product/store/product.spec.ts:895]: I'd imagine North America as a region having multiple currencies - CAD, USD. We'd lose out on specifying these if we remove currency_code right? Same case with EU as a region with multiple currencies.

From the looks of this PR, are we not merging the concepts of tax region and region here?

File: packages/medusa/src/api/store/products/helpers.ts [hunk 5]
```diff
@@ -30,3 +47,101 @@ export const maybeApplyStockLocationId = async (req: MedusaRequest, ctx) => {
...
+    return
+  }
+
+  const { isTaxInclusive } = req.taxContext.taxInclusivityContext
+  const taxService = req.scope.resolve(ModuleRegistrationName.TAX)
+
+  const taxRates = (await taxService.getTaxLines(
+    products.map(asTaxItem).flat(),
+    req.taxContext.taxLineContext
+  )) as unknown as ItemTaxLineDTO[]
```
Comment (riqwan) [inline packages/medusa/src/api/store/products/helpers.ts:75]: q: Isn't it our convention that we don't resolve modules outside of workflows?

File: packages/medusa/src/api/store/products/helpers.ts [hunk 6]
```diff
@@ -30,3 +47,101 @@ export const maybeApplyStockLocationId = async (req: MedusaRequest, ctx) => {
...
+
+  const taxRatesAsMap = new Map<string, ItemTaxLineDTO[]>()
+  taxRates.forEach((taxRate) => {
+    if (!taxRatesAsMap.has(taxRate.line_item_id)) {
+      taxRatesAsMap.set(taxRate.line_item_id, [])
+    }
+
+    taxRatesAsMap.get(taxRate.line_item_id)?.push(taxRate)
+  })
```
Comment (riqwan) [inline packages/medusa/src/api/store/products/helpers.ts]: Can just be called taxRatesMap

```suggestion
  const taxRatesMap = new Map<string, ItemTaxLineDTO[]>()
  
  taxRates.forEach((taxRate) => {
    if (!taxRatesMap.has(taxRate.line_item_id)) {
      taxRatesMap.set(taxRate.line_item_id, [])
    }

    taxRatesMap.get(taxRate.line_item_id)?.push(taxRate)
  })
```

File: packages/medusa/src/api/store/products/helpers.ts [hunk 7]
```diff
@@ -30,3 +47,101 @@ export const maybeApplyStockLocationId = async (req: MedusaRequest, ctx) => {
...
+  products.forEach((product) => {
+    product.variants?.forEach((variant) => {
+      const taxRatesForVariant = taxRatesAsMap.get(variant.id) || []
+      const { priceWithTax, priceWithoutTax } = getTaxPricesForVariant(
+        variant,
+        taxRatesForVariant,
+        isTaxInclusive
+      )
+
+      ;(variant.calculated_price as any).calculated_amount_with_tax =
```
Comment (riqwan) [inline packages/medusa/src/api/store/products/helpers.ts]: sugg: `calculated_price` is not guaranteed here, we might want to return early, earlier in the function if thats the case.

File: packages/medusa/src/api/utils/middlewares/products/set-tax-context.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  if (!region) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Region with id ${req.filterableFields.region_id} not found when populating the tax context`
+    )
+  }
+
+  // TODO: Add is_tax_inclusive here from pricing module
+  return {
+    isTaxInclusive: true,
```
Comment (olivermrbl) [inline packages/medusa/src/api/utils/middlewares/products/set-tax-context.ts]: question: Just for clarity, how would this flag get populated in a future iteration?

File: packages/medusa/src/api/store/products/helpers.ts [hunk 9]
```diff
@@ -30,3 +47,101 @@ export const maybeApplyStockLocationId = async (req: MedusaRequest, ctx) => {
...
+      if (!variant.calculated_price) {
+        return
+      }
+
+      return {
+        id: variant.id,
+        product_id: product.id,
+        product_name: product.title,
+        product_categories: product.categories?.map((c) => c.name),
+        product_category_id: product.categories?.[0]?.id,
```
Comment (riqwan) [inline packages/medusa/src/api/store/products/helpers.ts]: q: this looks weird here, especially the decision to choose the first category. Shouldn't it be `product_category_ids: [...]?`

File: packages/medusa/src/api/utils/middlewares/products/normalize-data-for-context.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+import { refetchEntities, refetchEntity } from "../../refetch-entity"
+
+export function normalizeDataForContext() {
+  return async (req: AuthenticatedMedusaRequest, _, next: NextFunction) => {
+    // If the product pricing is not requested, we don't need region information
+    let withCalculatedPrice = req.remoteQueryConfig.fields.some((field) =>
+      field.startsWith("variants.calculated_price")
+    )
+
+    // If the region is passed, we calculate the prices without requesting them.
```
Comment (olivermrbl) [inline packages/medusa/src/api/utils/middlewares/products/normalize-data-for-context.ts:13]: Why do we do that?

File: packages/medusa/src/api/utils/middlewares/common/clear-context-fields.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import { NextFunction } from "express"
+import { MedusaRequest } from "../../../../types/routing"
+
+export function clearContextFields(fields: string[]) {
```
Comment (riqwan) [inline packages/medusa/src/api/utils/middlewares/common/clear-context-fields.ts]: suggestion:

```suggestion
export function clearFiltersByKey(keys: string[]) {
```

File: packages/medusa/src/api/store/products/helpers.ts [hunk 12]
```diff
@@ -30,3 +47,101 @@ export const maybeApplyStockLocationId = async (req: MedusaRequest, ctx) => {
...
+    includesTax: isTaxInclusive,
+  })
+
+  const priceWithoutTax = isTaxInclusive
+    ? tax.minus(variant.calculated_price!.calculated_amount!).abs().toNumber()
+    : variant.calculated_price!.calculated_amount
+
+  const priceWithTax = isTaxInclusive
+    ? variant.calculated_price!.calculated_amount
+    : tax.plus(variant.calculated_price!.calculated_amount!).abs().toNumber()
```
Comment (riqwan) [inline packages/medusa/src/api/store/products/helpers.ts]: sugg: This should come from a generic helper, maybe even the `calculateTaxTotal` to return an object with these parts.

We'll be doing this in multiple places, might be good to consolidate them

File: packages/medusa/src/api/utils/middlewares/products/normalize-data-for-context.ts (hunk 10 above)
Comment (riqwan) [reply packages/medusa/src/api/utils/middlewares/products/normalize-data-for-context.ts:13]: Context - [CORE-2250](https://linear.app/medusajs/issue/CORE-2250)

File: packages/medusa/src/api/utils/middlewares/products/normalize-data-for-context.ts [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      }
+    }
+
+    // Finally, try to get it from the store defaults if not available
+    if (!regionId) {
+      const stores = await refetchEntities("store", {}, req.scope, [
+        "default_region_id",
+      ])
+      regionId = stores[0]?.default_region_id
+    }
```
Comment (riqwan) [inline packages/medusa/src/api/utils/middlewares/products/normalize-data-for-context.ts:55]: sugg: If region_id is not found from the cart or the query param, we should either throw or return without prices. If both the cart or query fail, we're already in a weird place. Picking from the store might put the query/customer in a weirder place if its a multi region store 😄 

I prefer the throw because I'd rather know an error is popping up as a dev so that I can fix it than silently returning.

**Review by riqwan — COMMENTED** [review]
Will write some general observations on the slack thread thats confusing me in this flow tomorrow. 🤞🏻

File: packages/medusa/src/api/store/products/helpers.ts [hunk 14]
```diff
@@ -30,3 +47,101 @@ export const maybeApplyStockLocationId = async (req: MedusaRequest, ctx) => {
 
   return entities.map((entity) => entity.stock_location_id)
 }
+
+export const wrapProductsWithTaxPrices = async <T>(
```
Comment (olivermrbl) [inline packages/medusa/src/api/store/products/helpers.ts:53]: We discussed the middleware approach briefly with @carlos-r-l-rodrigues yesterday evening, and he had a few good thoughts on this. Sharing here this morning, so we can kick off the discussion before his AM:

- Performance for store product list and retrieve endpoints is critical, so we should be careful about introducing too many operations at the HTTP layer that would harm it. Tax calculation has the potential to be in this category of operations. We want to use the search module as much as possible for the performance-critical endpoints.
- Performing updates on prices at the HTTP layer will make the sorting capabilities in the search module (when introduced) less useful. You can't rely on the sorting from the search engine because you are modifying your prices at a higher level in the stack.

I think both are valid points and good to open a discussion around. 

Right now (without the search module), there are at least two paths to go down:
- Leave the endpoint implementation as is, knowing we will have to rethink (and refactor) it when introducing the search module
- Put product variant price calculations into a separate endpoint, e.g. `GET /store/products/prices?id[]=...` (suggested by @carlos-r-l-rodrigues). This would mean a product page and product grid would to perform two requests to get the commonly required information to build such pages:
```
GET /store/products
GET /store/products/prices?id[]=...
```

Maybe there are other approaches to consider. Feel free to pitch in.

I don't have any strong opinion on this, there are pros and cons to both approaches. However, I do see the benefit of having a separate endpoint for price calculations in case we want to introduce more dynamic stuff in that context. Another case to consider in the future is inventory which will need to be computed and decorated dynamically on every request as well. We should try to follow a similar approach depending on what we decide here.

-----------------------

**On the search module**

We can index anything we want in the search module, but in the first iteration, it relies on the entity module events and doesn't offer much flexibility (correct me if I am wrong, @adrien2p, @carlos-r-l-rodrigues). Further down the line, we could imagine that the engine would expose "pre-indexing hooks," allowing us to perform calculations on resources before they are stored in the engine DB. 

An example use case of such pre-indexing hook would be price calculations. We could calculate product variant prices for each region and currency, so queries like the one we are discussing would require only a single request to the search engine.

File: integration-tests/http/__tests__/product/store/product.spec.ts (hunk 4 above)
Comment (sradevski) [reply integration-tests/http/__tests__/product/store/product.spec.ts:895]: @riqwan currently a region is tied to a single currency, so you would have to create two regions if you want to support checkout in multiple currencies. I already discussed this with Seb, and I agree with you, it is a limitation that we have in the system that we might want to address, but until then it doesn't make sense to accept currency code.

File: packages/medusa/src/api/store/products/helpers.ts (hunk 5 above)
Comment (sradevski) [reply packages/medusa/src/api/store/products/helpers.ts:75]: we are calling a non-standard endpoint here, so we don't have another way of doing it.

File: packages/medusa/src/api/store/products/helpers.ts (hunk 9 above)
Comment (sradevski) [reply packages/medusa/src/api/store/products/helpers.ts]: Yes I agree, and I think the tax module needs a rework, but it is outside the scope of this PR

File: packages/medusa/src/api/utils/middlewares/products/normalize-data-for-context.ts (hunk 13 above)
Comment (sradevski) [reply packages/medusa/src/api/utils/middlewares/products/normalize-data-for-context.ts:55]: If we do that, then the store shouldn't have a default region at all, do we agree on that? Let's see what others think about this.

File: packages/medusa/src/api/store/products/helpers.ts (hunk 14 above)
Comment (sradevski) [reply packages/medusa/src/api/store/products/helpers.ts:53]: @olivermrbl I agree on having the products endpoint very performant. In this case, unless you ask for the calculation by passing a cart or country context (we can even be more explicit), no calculations will happen, so it won't have an impact unless you need that information. 

Adding a separate endpoint won't solve anything, as it will have the exact same problems as doing `fields=*variants.calculated_price`. Doing that will actually introduce more issues
1. It will make the interface more complicated. 
2. The FE will need to do a join between products and prices
3. Sorting by price is a very common action, and that won't work on the `store/products` endpoint anymore.

If we want this tax calculation, we either do it like this, or we need to move that calculation to the pricing module. However, that won't reduce the number of operations, as you still need to list the tax lines and pass them to the pricing module.

Basically, our current architecture doesn't allow for significant performance improvements at the core. The main performance gains will have to come from a good caching layer.

File: packages/medusa/src/api/utils/middlewares/products/set-tax-context.ts (hunk 8 above)
Comment (sradevski) [reply packages/medusa/src/api/utils/middlewares/products/set-tax-context.ts]: @olivermrbl as I posted in the tax inclusive channel, we will store this as part of the pricing module (to begin with, set with either a currency or region rule), and we will just fetch it here before returning. The rest of the logic should remain unchanged (this will happen in a separate PR)

File: packages/medusa/src/api/store/products/helpers.ts (hunk 3 above)
Comment (sradevski) [reply packages/medusa/src/api/store/products/helpers.ts]: Yeah it is more readable, it's just that calculated amount is not a big number here, so instead of converting I did this. Let me revisit it though, I agree with you

File: packages/medusa/src/api/store/products/helpers.ts (hunk 1 above)
Comment (sradevski) [reply packages/medusa/src/api/store/products/helpers.ts]: Yes, I just need to check why the rate is nullable in the tax module, that's what the cast is for (and the TODO) 😄

File: packages/medusa/src/api/store/products/helpers.ts (hunk 14 above)
Comment (olivermrbl) [reply packages/medusa/src/api/store/products/helpers.ts:53]: If the resources are indexed in the search module, the request should be (still to be tested) highly performant and support sorting across modules, e.g. sort product variant prices. This is one of its main purposes. If we can index the prices with taxes taken into account, I think that would be ideal and probably what we should aim for to solve this. The two approaches I outlined above are both temporary imo., I just want us to consider which one we feel most comfortable with in the near term. There is no pretty way to do this kind of stuff right now, which is why we have all of this middleware acrobatics going on. 

I agree we will need to introduce heavy usage of caching, but I think that's a separate topic in this context and something we will need to do irregardless of this.

Comment (olivermrbl) [reply packages/medusa/src/api/store/products/helpers.ts:53]: I think the idea with the separate endpoint was also to have a place to introduce even more dynamic calculations in the future, in case we need that. E.g. dynamic to the point of being real-time, which wouldn't be possible to index in the search engine. Will let @carlos-r-l-rodrigues pitch in here.

File: packages/medusa/src/api/store/products/helpers.ts (hunk 1 above)
Comment (olivermrbl) [reply packages/medusa/src/api/store/products/helpers.ts]: Ah I see, makes sense 👍

File: packages/medusa/src/api/utils/middlewares/products/set-tax-context.ts (hunk 8 above)
Comment (olivermrbl) [reply packages/medusa/src/api/utils/middlewares/products/set-tax-context.ts]: As a separate data model in the pricing module right?

Comment (sradevski) [reply packages/medusa/src/api/utils/middlewares/products/set-tax-context.ts]: Yes exactly, that is also rule-based (at least that was the last thing we discussed with Seb)

Comment (olivermrbl) [reply packages/medusa/src/api/utils/middlewares/products/set-tax-context.ts]: Makes sense

File: packages/medusa/src/api/store/products/helpers.ts (hunk 14 above)
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api/store/products/helpers.ts:53]: Can we transform this into a workflow to fetch products?
This does not seem to be the case of middlewares, this is a cross domain calculation. It is more business logic than anything else at this point.

Comment (sradevski) [reply packages/medusa/src/api/store/products/helpers.ts:53]: I know where you are coming from, and I agree with having most of the BL outside the HTTP layer, but I don't think putting it in a workflow is the right solution here (or for any read-only operations). What we could do is create a simple helper function that wraps all of this logic and put it higher up the stack, WDYT?

File: packages/medusa/src/api/store/products/helpers.ts [hunk 15]
```diff
@@ -30,3 +49,81 @@ export const maybeApplyStockLocationId = async (req: MedusaRequest, ctx) => {
...
+        return
+      }
+
+      return {
+        id: variant.id,
+        product_id: product.id,
+        product_name: product.title,
+        product_categories: product.categories?.map((c) => c.name),
+        // TODO: It is strange that we only accept a single category, revisit the tax module implementation
+        product_category_id: product.categories?.[0]?.id,
```
Comment (olivermrbl) [inline packages/medusa/src/api/store/products/helpers.ts:119]: comment: I don't even think we support categories in the tax module right now. Last I remember the rules are fixed to a few props, but maybe this has changed

File: packages/medusa/src/api/utils/middlewares/products/normalize-data-for-context.ts (hunk 13 above)
Comment (olivermrbl) [reply packages/medusa/src/api/utils/middlewares/products/normalize-data-for-context.ts:55]: > If we do that, then the store shouldn't have a default region at all

Couldn't there be other use cases for a default region on the store? I am not it was added for the sole purpose of having a fallback for price calculations

**Review by olivermrbl — APPROVED** [review]
Overall LGTM. There's a lot going on in the HTTP layer, but it's hard to get around right now, so I am down for mergin as is

**Comment by sradevski** [comment]
@olivermrbl yeah agreed, I think we need to decide how routes are constructed and where such logic lies (we have several places where the middlewares and handlers are quite heavy code-wise), but it can be an independent cleanup task
